File: blind_75_leetcode/python/s0211_design_add_and_search_words_data_structure.py
from __future__ import annotations
from typing import List, Optional
import unittest


class WordDictionary:
    is_word: bool
    children: List[Optional[WordDictionary]]

    def __init__(self):
        self.is_word = False
        self.children = [None] * 26

    def addWord(self, word: str) -> None:
        node: WordDictionary = self
        for index in (ord(c) - ord("a") for c in word):
            if node.children[index] is None:
                node.children[index] = WordDictionary()
            node = node.children[index]  # type: ignore
        node.is_word = True

    def _search(self, word: str, index: int) -> bool:
        if index == len(word):
            return self.is_word
        elif word[index] == ".":
            return any(
                (
                    child._search(word, index + 1)
                    for child in self.children
                    if child is not None
                )
            )
        else:
            child = self.children[ord(word[index]) - ord("a")]
            if child is None:
                return False
            else:
                return child._search(word, index + 1)

    def search(self, word: str) -> bool:
        return self._search(word, 0)


class TestS211(unittest.TestCase):
    def test_works(self):
        w = WordDictionary()
        w.addWord("bad")
        w.addWord("dad")
        w.addWord("mad")
        self.assertEqual(w.search("pad"), False)
        self.assertEqual(w.search("bad"), True)
        self.assertEqual(w.search(".ad"), True)
        self.assertEqual(w.search("b.."), True)

File: blind_75_leetcode/src/solution/s0211_design_add_and_search_words_data_structure.rs
#![allow(dead_code)]
pub struct Solution {}

impl Solution {
    // This is a template
}

#[derive(Default)]
struct WordDictionary {
    is_word: bool,
    children: [Option<Box<WordDictionary>>; 26],
}

impl WordDictionary {
    fn new() -> Self {
        Self::default()
    }

    fn add_word(&mut self, word: String) {
        let mut node = self;
        for index in word.bytes().map(|b| (b - b'a') as usize) {
            if node.children[index].is_none() {
                node.children[index] = Some(Box::new(WordDictionary::new()));
            }
            node = node.children[index].as_mut().unwrap();
        }
        node.is_word = true;
    }

    fn _search(&self, bytes: &[usize], index: usize) -> bool {
        if index >= bytes.len() {
            self.is_word
        } else if bytes[index] == 27 {
            self.children
                .iter()
                .filter_map(|c| c.as_ref())
                .any(|c| c._search(bytes, index + 1))
        } else if let Some(node) = &self.children[bytes[index]] {
            node._search(bytes, index + 1)
        } else {
            false
        }
    }

    fn search(&self, word: String) -> bool {
        let bytes: Vec<_> = word
            .bytes()
            .map(|b| if b == b'.' { 27 } else { (b - b'a') as usize })
            .collect();
        self._search(&bytes, 0)
    }
}

#[cfg(test)]
mod tests_211 {
    use super::*;

    #[test]
    fn it_works() {
        let mut w = WordDictionary::new();
        w.add_word("bad".to_string());
        w.add_word("dad".to_string());
        w.add_word("mad".to_string());
        assert_eq!(w.search("b".to_string()), false);
        assert_eq!(w.search(".bad".to_string()), false);
        assert_eq!(w.search("pad".to_string()), false);
        assert_eq!(w.search("bad".to_string()), true);
        assert_eq!(w.search(".ad".to_string()), true);
        assert_eq!(w.search("b..".to_string()), true);
    }
}
